libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -2125,6 +2180,114 @@ bool AP_Mission::contains_item(MAV_CMD command) const
+void AP_Mission::reset_wp_history(void)
+{
+    for (uint8_t i = 0; i<AP_MISSION_MAX_WP_HISTORY; i++) {
+        _wp_index_history[i] = AP_MISSION_CMD_INDEX_NONE;
+    }
+    _resume_cmd.index = AP_MISSION_CMD_INDEX_NONE;
+    _flags.resuming_mission = false;
+    _repeat_dist = 0;
+}
+
+// store the latest reported position incase of mission exit and rewind resume
+void AP_Mission::update_exit_position(void)
+{
+    if (!AP::ahrs().get_position(_exit_position)) {
+        _exit_position.lat = 0;
+        _exit_position.lng = 0;
+    }
+}
+
+// calculate the location of the _resume_cmd wp and set as current
+bool AP_Mission::calc_rewind_pos(Mission_Command& rewind_cmd)
+{
+    // check for a recent history
+    if (_wp_index_history[LAST_WP_PASSED] == AP_MISSION_CMD_INDEX_NONE) {
+        // no saved history so can't rewind
+        return false;
+    }
+
+    // check that we have a valid exit position
+    if (_exit_position.lat == 0 && _exit_position.lng == 0) {
+        return false;
+    }
+
+    Mission_Command temp_cmd;
+    float rewind_distance = _repeat_dist; //(m)
+    uint16_t resume_index;
+    Location prev_loc = _exit_position;
+
+    for (int8_t i = (LAST_WP_PASSED); i>=0; i--) {
+
+        // to get this far there has to be at least one 'passed wp' stored in history.  This is to check incase 
+        // of history array no being completely filled with valid waypoints upon resume.
+        if (_wp_index_history[i] == AP_MISSION_CMD_INDEX_NONE) {
+            // no more stored history
+            resume_index = i+1;
+            break;
+        }
+
+        if (!read_cmd_from_storage(_wp_index_history[i], temp_cmd)) {
+            // if read from storage failed then don't use rewind
+            return false;
+        }
+
+        // calculate distance
+        float disttemp = prev_loc.get_distance(temp_cmd.content.location); //(m)
+        rewind_distance -= disttemp;
+        resume_index = i;
+
+        if (rewind_distance <= 0.0f) {
+            // history rewound enough distance to meet _repeat_dist requirement
+            rewind_cmd = temp_cmd;
+            break;
+        }
+
+        // store wp location as previous
+        prev_loc = temp_cmd.content.location;
+    }
+
+    if (rewind_distance > 0.0f) {
+        // then the history array was rewound all of the way without finding a wp distance > _repeat_dist distance.
+        // the last read temp_cmd will be the furthest cmd back in the history array so resume to that.
+        rewind_cmd = temp_cmd;
+        return true;
+    }
+
+    // if we have got this far the desired rewind distance lies between two waypoints stored in history array.
+    // calculate the location for the mission to resume
+
+    // the last wp read from storage is the wp that is before the resume wp in the mission order
+    Location passed_wp_loc = temp_cmd.content.location;
+
+    // fetch next destination wp
+    if (!read_cmd_from_storage(_wp_index_history[resume_index+1], temp_cmd)) {
```

Comment:
resume_index needs to be bound checked here. It has the potential of being LAST_WP_PASSED+1 and then used as LAST_WP_PASSED +1 + 1 here which is out of range.

Reply:
It's not possible for the resume index to be out of range here.  LAST_WP_PASSED = length of the array - 2.